### PR TITLE
Enable solving QP in parallel with ProxQP 

### PR DESCRIPTION
**Pull request by Bambade**
Enable solving a vector of QPs in parallel using open mp with either the sparse or dense ProxQP backend.

The code is unit tested in both C++ and python.

### CODE REVIEW

**Comment by jcarpent** [comment]
@fabinsch I think we should disable the parallel feature for PiPy. As for Pinocchio, we should be able to deactivate the parallel module.

**Comment by jcarpent** [comment]
@nim65s Do you know how to integrate OpenMP to PyPI?

**Comment by nim65s** [comment]
I have no idea. I'll give it a try this summer.

File: bindings/python/src/expose-all.cpp [hunk 1]
```diff
@@ -20,6 +20,11 @@ exposeCommon(pybind11::module_ m)
 {
   exposeResults<T>(m);
   exposeSettings<T>(m);
+#ifdef PROXSUITE_PYTHON_INTERFACE_WITH_OPENMP
+  m.def("omp_get_max_threads",
+        &omp_get_max_threads,
+        "Returns an upper bound on the number of threads that could be used.");
```
Comment (jcarpent) [inline bindings/python/src/expose-all.cpp:26]:
```suggestion
        "Returns the max number of threads that could be used by OpenMP.");
```

File: bindings/python/src/expose-qpvector.hpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
+//
+// Copyright (c) 2022 INRIA
```
Comment (jcarpent) [inline bindings/python/src/expose-qpvector.hpp:2]:
```suggestion
// Copyright (c) 2023 INRIA
```

File: include/proxsuite/proxqp/parallel/qp_solve.hpp [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+void
+qp_solve_in_parallel(optional<const size_t> num_threads,
+                     std::vector<proxqp::dense::QP<T>>& qps)
+{
+  if (num_threads != nullopt) {
+    set_default_omp_options(num_threads.value());
+  } else {
+    size_t NUM_THREADS =
+      std::max((size_t)(omp_get_max_threads() / 2 - 2), (size_t)(1));
+    set_default_omp_options(NUM_THREADS);
```
Comment (jcarpent) [inline include/proxsuite/proxqp/parallel/qp_solve.hpp:26]: Why?

File: include/proxsuite/proxqp/dense/wrapper.hpp [hunk 4]
```diff
@@ -484,6 +484,57 @@ operator!=(const QP<T>& qp1, const QP<T>& qp2)
   return !(qp1 == qp2);
 }
 
+///// BatchQP object
+template<typename T>
+struct VectorQP
```
[the remark this answers is not in the document]
Comment (jcarpent) [reply include/proxsuite/proxqp/dense/wrapper.hpp:489]: In any case, I do think the name Batch is better suited.

File: include/proxsuite/proxqp/dense/wrapper.hpp [hunk 5]
```diff
@@ -484,6 +484,57 @@ operator!=(const QP<T>& qp1, const QP<T>& qp2)
...
+  dense::isize m_size;
+
+  VectorQP(long unsigned int batchSize)
+  {
+    if (vector_qp.max_size() != batchSize) {
+      vector_qp.clear();
+      vector_qp.reserve(batchSize);
+    }
+    m_size = 0;
+  }
```
Comment (jcarpent) [inline include/proxsuite/proxqp/dense/wrapper.hpp:505]:
```suggestion
  {
    if (vector_qp.max_size() != batch_size) {
      vector_qp.clear();
      vector_qp.reserve(batch_size);
    }
    m_size = 0;
  }
```

File: include/proxsuite/proxqp/dense/wrapper.hpp [hunk 6]
```diff
@@ -484,6 +484,57 @@ operator!=(const QP<T>& qp1, const QP<T>& qp2)
...
+    vector_qp.emplace_back(dim, n_eq, n_in);
+    auto& qp = vector_qp.back();
+    m_size++;
+    return qp;
+  };
+
+  /*!
+   * Inserts a qp to the end of vector_qp
+   */
+  void insert(QP<T>& qp) { vector_qp.emplace_back(qp); };
```
Comment (jcarpent) [inline include/proxsuite/proxqp/dense/wrapper.hpp:523]:
```suggestion
  void insert(QP<T>& qp) { vector_qp.emplace_back(qp); };
```

**Review by jcarpent — CHANGES_REQUESTED** [review]

File: include/proxsuite/proxqp/dense/wrapper.hpp (hunk 6 above)
Comment (fabinsch) [reply include/proxsuite/proxqp/dense/wrapper.hpp:523]: what is the difference here?

File: include/proxsuite/proxqp/dense/wrapper.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (jcarpent) [reply include/proxsuite/proxqp/dense/wrapper.hpp]: You cannot do a benchmark like that. You need to average over various calls. Please see Pinocchio for that.

File: test/src/parallel_qp_solve.cpp [hunk 7]
```diff
@@ -189,8 +245,8 @@ DOCTEST_TEST_CASE("test sparse vector QP")
   }
 
   const size_t NUM_THREADS = (size_t)omp_get_max_threads();
-  proxsuite::proxqp::parallel::qp_solve_in_parallel((size_t)(NUM_THREADS / 2),
-                                                    qps_vector);
+  proxsuite::proxqp::parallel::qp_solve_in_parallel(qps_vector,
```
Comment (jcarpent) [inline test/src/parallel_qp_solve.cpp:248]: Why do you have two namings different between cpp and python? (`qp_solve_in_parallel` and `solve_in_parallel`)

Comment (jcarpent) [reply test/src/parallel_qp_solve.cpp:248]: I don't think the namespace parallel is helpful. I would remove it. In addition, the Python bindings have an extra dense namespace. Why?

File: include/proxsuite/proxqp/parallel/qp_solve.hpp [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+qp_solve_in_parallel(std::vector<proxqp::dense::QP<T>>& qps,
+                     optional<const size_t> num_threads = nullopt)
+{
+  if (num_threads != nullopt) {
+    set_default_omp_options(num_threads.value());
+  } else {
+    size_t NUM_THREADS =
+      std::max((size_t)(omp_get_max_threads() / 2), (size_t)(1));
+    set_default_omp_options(NUM_THREADS);
+  }
```
Comment (jcarpent) [inline include/proxsuite/proxqp/parallel/qp_solve.hpp:27]:
```suggestion
  size_t NUM_THREADS =
      std::max((size_t)(omp_get_max_threads() / 2), (size_t)(1)); // TODO(jcarpent): find optimal dispatch
  if (num_threads != nullopt) {
    NUM_THREADS = num_threads.value();
  } 
  set_default_omp_options(NUM_THREADS);
```

File: include/proxsuite/proxqp/parallel/qp_solve.hpp [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+qp_solve_in_parallel(proxqp::dense::BatchQP<T>& qps,
+                     optional<const size_t> num_threads = nullopt)
+{
+  if (num_threads != nullopt) {
+    set_default_omp_options(num_threads.value());
+  } else {
+    size_t NUM_THREADS =
+      std::max((size_t)(omp_get_max_threads() / 2), (size_t)(1));
+    set_default_omp_options(NUM_THREADS);
+  }
```
Comment (jcarpent) [inline include/proxsuite/proxqp/parallel/qp_solve.hpp:49]: @fabinsch COuld you apply the same strategy here?

Comment (fabinsch) [reply include/proxsuite/proxqp/parallel/qp_solve.hpp:49]: yes sure

File: include/proxsuite/proxqp/dense/wrapper.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (fabinsch) [reply include/proxsuite/proxqp/dense/wrapper.hpp]: okay, I updated the benchmark.

File: test/src/parallel_qp_solve.cpp (hunk 7 above)
Comment (fabinsch) [reply test/src/parallel_qp_solve.cpp]: I renamed both to `solve_in_parallel` and made in consistent between cpp and python.

**Review by jcarpent — DISMISSED** [review]

File: include/proxsuite/proxqp/dense/wrapper.hpp (hunk 6 above)
Comment (jcarpent) [reply include/proxsuite/proxqp/dense/wrapper.hpp:523]: None

**Pull request merged**